edb/server/conn_pool/src/conn.rs

```diff
@@ -0,0 +1,348 @@
+    fn reconnect(
+        &self,
+        conn: Self::Conn,
+        db: &str,
+    ) -> impl Future<Output = ConnResult<Self::Conn>> + 'static;
+
+    /// Perform a graceful disconnect operation on the given connection.
+    fn disconnect(&self, conn: Self::Conn) -> impl Future<Output = ConnResult<()>> + 'static;
+}
+
+#[derive(Debug)]
+pub struct Conn<C: Connector> {
+    inner: Rc<RefCell<ConnInner<C>>>,
+}
+
+impl<C: Connector> PartialEq for Conn<C> {
+    fn eq(&self, other: &Self) -> bool {
+        Rc::ptr_eq(&self.inner, &other.inner)
+    }
+}
+
+impl<C: Connector> Eq for Conn<C> {}
+
+impl<C: Connector> Clone for Conn<C> {
+    fn clone(&self) -> Self {
+        Self {
+            inner: self.inner.clone(),
+        }
+    }
+}
+
+impl<C: Connector> Conn<C> {
+    pub fn new(
+        f: impl Future<Output = ConnResult<C::Conn>> + 'static,
+        metrics: &MetricsAccum,
+    ) -> Self {
+        metrics.insert(MetricVariant::Connecting);
+        Self {
+            inner: Rc::new(RefCell::new(ConnInner::Connecting(
+                Instant::now(),
+                f.boxed_local(),
+            ))),
+        }
+    }
+
+    #[inline(always)]
+    pub fn with_handle<T>(&self, f: impl Fn(&C::Conn) -> T) -> Option<T> {
+        match &*self.inner.borrow() {
+            ConnInner::Active(_, conn, ..) => Some(f(conn)),
+            _ => None,
+        }
+    }
+
+    #[inline]
+    fn transition(&self, f: impl FnOnce(ConnInner<C>) -> ConnInner<C>) {
+        let mut lock = self.inner.borrow_mut();
+        let inner = std::mem::replace(&mut *lock, ConnInner::Transition);
+        *lock = f(inner);
+    }
+
+    pub fn close(&self, connector: &C, metrics: &MetricsAccum) {
+        self.transition(|inner| match inner {
+            ConnInner::Idle(t, conn, ..) => {
+                metrics.transition(
+                    MetricVariant::Idle,
+                    MetricVariant::Disconnecting,
+                    t.elapsed(),
+                );
+                let f = connector.disconnect(conn).boxed_local();
+                ConnInner::Disconnecting(Instant::now(), f)
+            }
+            _ => unreachable!(),
+        });
+    }
+
+    pub fn discard(&self, connector: &C, metrics: &MetricsAccum) {
+        self.transition(|inner| match inner {
+            ConnInner::Active(t, conn, ..) => {
+                metrics.transition(
+                    MetricVariant::Active,
+                    MetricVariant::Disconnecting,
+                    t.elapsed(),
+                );
+                let f = connector.disconnect(conn).boxed_local();
+                ConnInner::Disconnecting(Instant::now(), f)
+            }
+            _ => unreachable!(),
+        });
+    }
+
+    pub fn transfer(&self, connector: &C, from: &MetricsAccum, to: &MetricsAccum, db: &str) {
+        self.untrack(from);
+        self.transition(|inner| match inner {
+            ConnInner::Idle(_t, conn, ..) | ConnInner::Active(_t, conn, ..) => {
```

Comment:
`transfer` is used for two different cases -- basically this handles the case of "stealing" a connection on release (when it would still be marked as active) vs "stealing" an idle connection.